File: Python/Codes/Practice/LeetCode/539 Minimum Time Difference.py

class Solution(object):
    def findMinDifference(self, timePoints):
        """
        :type timePoints: List[str]
        :rtype: int
        """

        timePoints.sort()
        miniDelta = 24*60
        length = len(timePoints)
        for i in range(length):
            delta = abs(self.diffMinute(timePoints[(i-1+length) % length],timePoints[i]))
            delta = min(delta, 24*60 - delta);
            if delta < miniDelta:
                miniDelta = delta
        
        return miniDelta

    def diffMinute(self, timeA, timeB):
        A = timeA.split(':')
        B = timeB.split(':')
        return (int(B[0]) - int(A[0]))*60 + (int(B[1]) - int(A[1]))


# Test Case 
import unittest
class SolutionTest(unittest.TestCase):
    def test_all(self):
        s = Solution()
        self.assertEqual(s.findMinDifference(["23:59", "00:00"]),1)
        self.assertEqual(s.findMinDifference(["00:00","23:59", "00:00"]),0)
        self.assertEqual(s.findMinDifference(["05:31", "22:08", "00:35"]),147)
        self.assertEqual(s.findMinDifference(["17:41","22:06","20:59","17:37","04:14"]),4)
        

if __name__ == '__main__':
    unittest.main()

File: Cpp/Codes/Practice/LeetCode/539 Minimum Time Difference.cpp
#include <gtest/gtest.h>
#include <algorithm>
#include <math.h>
#include <Common/util.h>
using namespace std;

class MinimumTimeDifference
{
public:
	int findMinDifference(vector<string>& timePoints) 
	{
		sort(timePoints.begin(), timePoints.end());
		int minDiff = numeric_limits<int>::max();
		int length = timePoints.size();
		for (int i = 0; i < length; ++i)
		{
			int diff = abs(
				diffMinute(
					timePoints[(i-1+length) % length],
					timePoints[i]));
			diff = min(diff, 24 * 60 - diff);
			if (diff < minDiff)
			{
				minDiff = diff;
			}
		}

		return minDiff;
	}

	int diffMinute(string A, string B)
	{
		char* timeA = (char*)A.c_str();
		char* timeB = (char*)B.c_str();
		timeA[2] = '\0';
		timeB[2] = '\0';

		return (atoi(timeB) - atoi(timeA)) * 60 +
			(atoi(timeB + 3) - atoi(timeA + 3));
	}
};

TEST(LeetCode, tMinimumTimeDifference)
{
	MinimumTimeDifference s;
	vector<string> timePoints;
	timePoints.push_back("23:59");
	timePoints.push_back("00:00");
	ASSERT_EQ(s.findMinDifference(timePoints), 1);

	timePoints.clear();
	timePoints.push_back("00:00");
	timePoints.push_back("23:59");
	timePoints.push_back("00:00");
	ASSERT_EQ(s.findMinDifference(timePoints), 0);

	timePoints.clear();
	timePoints.push_back("05:31");
	timePoints.push_back("22:08");
	timePoints.push_back("00:35");
	ASSERT_EQ(s.findMinDifference(timePoints), 147);

	timePoints.clear();
	timePoints.push_back("17:41");
	timePoints.push_back("22:06");
	timePoints.push_back("20:59");
	timePoints.push_back("17:37");
	timePoints.push_back("04:14");
	ASSERT_EQ(s.findMinDifference(timePoints), 4);
}